File: shareable-pedigree-link/python/src/shareable_pedigree_link/snippet_builder.py
"""Pure transform from (embed URL, label, key) to an HTML snippet."""

from __future__ import annotations

from dataclasses import dataclass
from html import escape

_IFRAME_HEIGHT_PX = 640


@dataclass(frozen=True)
class SnippetRequest:
    embed_url: str
    label: str
    minted_at: str
    plaintext_key: str
    revoke_url: str


def build_snippet(request: SnippetRequest) -> str:
    src = escape(request.embed_url, quote=True)
    title = escape(request.label, quote=True)
    minted_at = escape(request.minted_at, quote=False)
    return (
        f"<!-- Evagene embeddable pedigree \u2014 read-only key minted {minted_at} -->\n"
        f'<iframe src="{src}" title="{title}" '
        f'width="100%" height="{_IFRAME_HEIGHT_PX}" frameborder="0"></iframe>\n'
        "\n"
        f"Minted API key: {request.plaintext_key}   "
        f"(stored only here \u2014 revoke at {request.revoke_url})\n"
    )

File: shareable-pedigree-link/node/src/snippetBuilder.ts
export interface SnippetRequest {
  readonly embedUrl: string;
  readonly label: string;
  readonly mintedAt: string;
  readonly plaintextKey: string;
  readonly revokeUrl: string;
}

const IFRAME_HEIGHT_PX = 640;

export function buildSnippet(request: SnippetRequest): string {
  const src = escapeAttribute(request.embedUrl);
  const title = escapeAttribute(request.label);
  const mintedAt = escapeHtml(request.mintedAt);
  return (
    `<!-- Evagene embeddable pedigree \u2014 read-only key minted ${mintedAt} -->\n` +
    `<iframe src="${src}" title="${title}" width="100%" height="${IFRAME_HEIGHT_PX.toString()}" frameborder="0"></iframe>\n` +
    `\n` +
    `Minted API key: ${request.plaintextKey}   ` +
    `(stored only here \u2014 revoke at ${request.revokeUrl})\n`
  );
}

const ATTRIBUTE_ESCAPES: ReadonlyMap<string, string> = new Map([
  ['&', '&amp;'],
  ['<', '&lt;'],
  ['>', '&gt;'],
  ['"', '&quot;'],
]);

const HTML_ESCAPES: ReadonlyMap<string, string> = new Map([
  ['&', '&amp;'],
  ['<', '&lt;'],
  ['>', '&gt;'],
]);

function escapeAttribute(value: string): string {
  return replaceFrom(value, ATTRIBUTE_ESCAPES);
}

function escapeHtml(value: string): string {
  return replaceFrom(value, HTML_ESCAPES);
}

function replaceFrom(value: string, table: ReadonlyMap<string, string>): string {
  let result = '';
  for (const char of value) {
    result += table.get(char) ?? char;
  }
  return result;
}
